feat(robot): Support logging arbitrary data to the console

jhermsmeier opened the pull request:
This adds a new command recognition for message type "log",
to enable `console.log()`ing arbitrary data to the parent process console.

Not sure if we'd really want this in, but it's very handy for debugging,
and would keep us from writing these 3 lines every time
we want to inspect something in between those processes.

Change-Type: minor

lurch commented:
Is there, or should there be, documentation of what methods / messages / etc. the 'robot' understands / emits?

jviotti commented:
I like the idea. Lets add some documentation for this like @lurch pointed out. Maybe inside the README at `lib/child-writer`? We could also have a small utility function that creates and prints a "log" robot message, so we don't have to manually construct it.

lurch commented:
IMHO we should wait until after #1221 has been merged, and then rebase this one.

jviotti commented:
Yeah, sounds like a good way to avoid some more complex conflicts.

jviotti commented:
The linked PR was merged! We have some conflicts here, but they are hopefully very simple to deal with.

lurch commented on lib/child-writer/index.js:
```diff
@@ -139,6 +139,8 @@ exports.write = (image, drive, options) => {
       if (messageCommand === 'error') {
         emitError(robot.recomposeErrorMessage(parsedMessage));
+      } else if (messageCommand === 'log') {
+        console.log(message.data);
```
Should this be `messageData`? Or `messageData.data`? Or is `message.data` indeed correct??
I think I need to go and lay down, I'm getting dizzy ;-)

jhermsmeier replied:
Oh, missed that – fixed!

jviotti replied:
Yeah, this should be `messageData`.

lurch commented:
\*bump\* https://github.com/resin-io/etcher/pull/1212#issuecomment-288897255  ;-)

lurch commented:
re-ping @jhermsmeier - do you want to address @jviotti 's comments in https://github.com/resin-io/etcher/pull/1212#issuecomment-288897255

jhermsmeier commented:
> We could also have a small utility function that creates and prints a "log" robot message, so we don't have to manually construct it.

Added `robot.log({ data: 'datadata' })` for even fewer chars to type (which just calls `robot.printMessage('log', data)`).

> I like the idea. Lets add some documentation for this like @lurch pointed out. Maybe inside the README at lib/child-writer?

Added an example to `lib/shared/robot/README.md`, since the message passing is already documented there.

lurch commented on lib/shared/robot/index.js:
```diff
@@ -215,3 +215,17 @@ exports.printError = (error) => {
+ * @function
+ * @public
+ *
+ * @param {Object} [data] - data
```
Does it always need to be an Object, or would passing in a String work too?

jhermsmeier replied:
It does have to be an object, otherwise `buildMessage()` will yell (see line 58: https://github.com/resin-io/etcher/pull/1212/files#diff-0ccd387a7e7d9c117af87e3bdbd6a9ebR58)
